### PR TITLE
Ports TG Components and R&D Tweaks

### LINKED ISSUE #7515: Circuit Imprinter doesn't take metal, but requires metal to print certain boards
**Problem Description**:
Circuit Imprinter doesn't take metal, but requires metal to print certain boards

**What did you expect to happen**: 
The circuit imprinter should either not require metal, or should accept it when a recipe requires it.

**What happened instead**:
Several board recipes required metal. Upon trying to insert metal into the machine, it refused to take the metal.

**Why is this bad/What are the consequences:**
This issue prevents R&D from printing any circuit boards that require metal.

**Steps to reproduce the problem**:
Try to print a computer board for example, it requires metal. Try to insert metal into the machine. The machine will reject the metal.

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was
anything else out of the ordinary happening?)**: Anything else you can tell us.
Nothing out of the ordinary.

### LINKED ISSUE #7679: "wired network card" has impossible requirements
Wired Network Card that can be produced only in a circuit imprinter requires metal to be constructed.

Metal cannot be inserted into a circuit imprinter.

### LINKED ISSUE #7788: Unable to Place Metal into Circuit Imprinter
image

**Problem Description**:
You cannot place metal into the circuit printer.

**What did you expect to happen**: 
The ability to place the resources required for certain items from the imprinter into the imprinter.

**What happened instead**:
The inability to place metal into the imprinter.

**Why is this bad/What are the consequences:**
Certain items cannot be printed from the imprinter even if the research is available for such.

**Steps to reproduce the problem**:
• Acquire the research required for an item in the circuit imprinter that requires metal
• Be unable to place metal into the imprinter and therefore unable to print the item

### LINKED ISSUE #7887: Circtuir Imprinters and Metal
**Problem Description**:
There are some things you need metal to print from a circuit imprinter, but circuit imprinters cannot accept metal.

**What did you expect to happen**: 
I would expect to be able to print anything a circuit imprinter can create given the resources for it.

**What happened instead**:
I was unable to insert the require materials.

**Why is this bad/What are the consequences:**
It should be possible, given the materials and tech levels, to print anything a machine is capable of printing.

**Steps to reproduce the problem**:
Attempt to print any network card from a circuit imprinter.

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's component system.

Component wise, I will only be implementing the materials component at first.

- [x] Get components ported and working
- [x] Get materials component ported and working
- [x] Implements materials component where old material datum was used
- [x] Convert Circuit Imprinter over to using materials component
- [x] Convert Mech Fab over to using materials component
- [x] Refactor pod Fabricator to be a child of mech fabricator, instead of horrendous copy-pasta
- [x] Refactor Mint to utilize the material datum; also refactors money bags to be an actual storage type, instead of some horrifying snowflaked abomination
- [x] Refactor robot fabricator to be a child of mech fabricator, instead of horrendous copy-pasta

Components are a modular system for attaching behaviors and properties to objects. They're a very OOP way of adding universal behaviors to something without having to re-code/snowflake coded behaviors to an object over and over again. For example, say you want to give something the property of exploding into spectacular particles and generating an actual explosion when that object is deleted. You could have an entire code block that has this behavior then copy-paste it all over where it's needed, or you could have this behavior tied to a component, then set how/when it triggers.

A component has specific properties about it; when it's created, you can have it listen for "signals" and what to do when those signals trigger. As an example with the materials component, it listens for the object being attacked, then attempts to load what you're attacking it with into the object you're attacking and covert it from an object into metal/glass/plasma/diamond/etc. 

These signals can be anything and placed almost anywhere; it could be set up to occur when an object is attacked, created, destroyed, thrown, or what have you. 

Accessing a component is also fairly easy; once an object as a component added to it, you simply access it with `GET_COMPONENT`, then provide the arguments of what you're attempting to access on a component and the path of the component accessed, then you can begin interacting with it.

On first examination, the system can be a bit daunting and confusing, but once you wrap your head around it, the possibilities with components are great; it not only means more object oriented code, but it means adding/removing attributes to various objects on the fly and easily; instead of having hard coded objects with hard coded behaviors; we could end up with lots of base level objects with a myriad of components attached to them to govern that object's behavior.

other notes:

- Adds in `flag_2` (in a future PR, we'll implement `flag_1`; this is just to avoid future flag conflicts
- Adds future support for blocking holodeck items from being converted into materials (or utilized in other construction things)
- Adds support for blocking specific items from being converted into materials
- Refactored the Pod fabricator to be a child of the mech fabricator; functionality is the same; this just eliminates a LOT of copy-pasta
  - Moved the pod fab on BoxStation to be beside the disposal (as the wall blocks building with the new system)

Other things
- Added in the rest of the materials that Paradise doesn't have: bluespace mesh and plastic
- Plastic and bluespace crystals now have their appropriate materials
- Increased the plastic pack sheet amount to 50 and reduced cost to 10
- Can make waterbottles (large and small) and caution signs with sheets of plastic now
- Can create plastic sheets by combining oil, acid, and ash
- Removed the acid cost from most circuits
  - Following TG's lead on this one here; this really doesn't serve much of a purpose but needless busywork
- Added bluespace material requirement on a few parts: bluespace stock parts, phazon parts, bags of holding, and the likes all now require bluespace lattice.

Waterbottles sprites:

img

Fixes #7515 
Fixes #7679 
Fixes #7788 
Fixes #7887
Closes https://github.com/ParadiseSS13/Paradise/pull/8808

:cl: Fox McCloud
add: Can make water bottles and caution signs out of plastic sheets
tweak: Increased cargo plastic sheet count to 50 and reduced cost to 10
add: Adds a recipe for making plastic sheets: oil, acid, and ash
fix: Can now make a designs on the circuit imprinter that use metal and other exotic materials
tweak: Removed the acid cost from most circuit recipes
tweak:  bluespace material requirement on a few parts: bluespace stock parts, phazon parts, bags of holding, and the likes all now require bluespace lattice.
/:cl:

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Well, tested it out with the autolathe; it works perfectly well. the component garbage collects fine, as does the autolathe itself; couldn't see any particular bugs or anything.

File: code/datums/components/_component.dm [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	return
+
+/datum/component/proc/_GetInverseTypeList(our_type = type)
+	#if DM_VERSION >= 513
+	#warning 512 is definitely stable now, remove the old code
+	#endif
+
+	#if DM_VERSION < 512
+	//remove this when we use 512 full time
+	set invisibility = 101
```
Comment (Fox-McCloud) [inline code/datums/components/_component.dm]: This is actually needed if you're running less than 512, or components will just straight up not work and runtime like mad.

Comment (KasparoVy) [reply code/datums/components/_component.dm]: Voodoo

**Comment by Anticept** [comment]
Well then.

Here I thought I was being creative last year when I started roadmapping a science and industry redo using this same component concept heh

**Comment by Fox-McCloud** [comment]
Andd with the mech fab and circuit imprinters converted over, I think that should about do it for this PR.

**Comment by matt81093** [comment]
So.. this closes #8808 and.. you added some component stuff with the changes?

File: code/modules/research/designs/AI_module_designs.dm [hunk 2]
```diff
@@ -8,7 +8,7 @@
 	id = "freeform_module"
 	req_tech = list("programming" = 5, "materials" = 4)
 	build_type = IMPRINTER
-	materials = list(MAT_GLASS = 1000, "sacid" = 20, MAT_GOLD = 100)
+	materials = list(MAT_GLASS = 1000, MAT_GOLD = 100)
```
Comment (matt81093) [inline code/modules/research/designs/AI_module_designs.dm]: You removed the acid requirement entirely?

File: code/modules/research/designs/computer_part_designs.dm [hunk 3]
```diff
@@ -63,7 +63,8 @@
 	id = "netcard_basic"
 	req_tech = list("programming" = 2, "engineering" = 1)
 	build_type = IMPRINTER
-	materials = list(MAT_METAL = 250, MAT_GLASS = 100, "sacid" = 20)
+	materials = list(MAT_METAL = 250, MAT_GLASS = 100)
+	reagents_list = list("sacid" = 20)
```
Comment (matt81093) [inline code/modules/research/designs/computer_part_designs.dm]: Although you kept the acid requirement here.

File: code/modules/research/designs/AI_module_designs.dm (hunk 2 above)
Comment (Fox-McCloud) [reply code/modules/research/designs/AI_module_designs.dm]: Yes, as I documented in the post, I removed acid requirements from everything; for the most part, it's pointless busywork, and not an actual gate like, say, gold, diamond, or bluespace mesh is.

**Comment by Fox-McCloud** [comment]
The difference? One is a bit more object oriented.

When adding materials datum to something, you not only have to create the datum, but you have to specify *how* you interact with the materials datum. For instance, everywhere that there's a material datum that you insert something into, you have to re-code that behavior.

The material component has universal behaviors defined on the component itself; for example, it listens for the signal "attackby" and if it hears it, then it will attempt to let you insert materials into the owner of the component.

Obviously, there's unique behaviors that do still have to be specifically coded (like design cost and such), but universal behaviors (inserting, for instance) are all handled on the component now.

**Comment by taukausanake** [comment]
I really don't like those sprites. Can't just make new sprites for the new animations introduced here?

File: code/modules/research/designs/AI_module_designs.dm (hunk 2 above)
Comment (Fox-McCloud) [reply code/modules/research/designs/AI_module_designs.dm]: Addressed the inconsistency of these two parts; their base versions required acid, but their advanced versions did not; now, none of them do.

**Comment by Fox-McCloud** [comment]
Reverted the sprites for now.

**Pull request merged**